safe-setting v1 release (#137)
github/safe-settings: README.md

# GitHub Safe-Settings

[![Node.js CI](https://github.com/github/safe-settings/actions/workflows/node-ci.yml/badge.svg)](https://github.com/github/safe-settings/actions/workflows/node-ci.yml)
[![Dependabot][dependabot-badge]][dependabot-link]

`Safe-settings`– an app to manage policy-as-code and apply repository settings to repositories across an organization.

It is loosely based on [Settings Probot](https://github.com/probot/settings) GitHub App. However, it differs from the original probot settings app in several important ways:

1. In `safe-settings` all the settings are stored centrally in an `admin` repo within the organization. 
1. There are 3 levels at which the settings could be managed:
   1. Org-level settings are defined in `.github/settings.yml` 

>    1. The org-level settings could be overridden at a  `suborg` level. These would be a collection of repos belonging to a project or a business unit, or a team. The `suborg`settings reside in the `.github/suborgs`folder.

Reviewer: Can a repository belong to more than one suborg?

Author: Currently, it doesn't work well if a repo is on multiple suborgs because there is no precedence and the last suborg config for the repo will override the previous ones.

   1. Lastly, we could override settings for individual repos by creating a repo specific yaml in `.github/repos`folder
1. `Safe-settings` explicitly looks in the `admin` repo in the organization for the settings files. The `admin` repo should be a restricted repository with `branch protections` and `codeowners` . 
1. To address the scalability concerns for large organizations having thousands of repos, it is recommended to break the settings into org-level, suborg-level, and repo-level units. This will allow different teams to be define and manage policies for their specific projects or business units.

## Table of Contents

> [TOC]

Reviewer: Perhaps this should just be removed for now...

Author: Totally makes sense. The intent was it would be auto-generated but that didn;t pan out.

## How it works

The App listens to the following webhook events and does the following:

- **push**: If settings are created or settings are modified, that is, if  push happens in the `default` branch of the `admin` repo and the file added or changed is `.github/settings.yml` or `.github/repos/*.yml`or `.github/suborgs/*.yml`, then the settings would be applied either globally to all the repos, or specific repos. For each repo the settings that would be applied depend on the default settings for the org, overlayed with settings for the suborg that the repo belongs to, overlayed with the settings for that specific repo.
  

> - **repository.created**: If a repository is created in the org, the settings for the repo - the default settings for the org, overlayed with settings for the suborg that the repo belongs to, overlayed with the settings for that specific repo - is applied. To prevent `infinite loop` if the repo created event is trigged by itself(bot), it would be ignored. This ensures that the policy is enforced even to nee repositories.

Reviewer: I'm not sure I understand the `infinite loop` issue here.
```suggestion
- **repository.created**: If a repository is created in the org, the settings for the repo - the default settings for the org, overlayed with settings for the suborg that the repo belongs to, overlayed with the settings for that specific repo - is applied. To prevent `infinite loop` if the repo created event is trigged by itself(bot), it would be ignored. This ensures that the policy is enforced even to a single repositories.
```

Reviewer: Can a file in `.github/repos/` be create prior to the creation of a repository to which it will apply without error?

> - **repository.edited**: If the `default` branch is changed if the the settings that would be applied for the repo

Reviewer: I'm not sure what this is trying to state.

> **Exceptions:** Certain repos may be exempted by specifying them in a runtime config file called `depolyment-settings.yml`. If no file is specified, then the following repositories -  `'admin', '.github', 'safe-settings'` are exempted by default.

Reviewer: Would it be correct to state the location as the following:
```suggestion
**Exceptions:** Certain repos may be exempted by specifying them in a runtime config file called `.github/depolyment-settings.yml`. If no file is specified, then the following repositories -  `'admin', '.github', 'safe-settings'` are exempted by default.
```

Reviewer: Given the generic name `admin` for the repo I would propose any/all configurations be more clearly related to safe-settings through subdirectory usage.

Author: > Would it be correct to state the location as the following:

The deployment-settings.yml is kept at the root directory where the safe-settings is running.

Author: > Given the generic name `admin` for the repo I would propose any/all configurations be more clearly related to safe-settings through subdirectory usage.

Let us discuss this. I am interested in knowing more about this proposal.

1. The `teams` section contains the list of teams that need to be added to all the repositories in the org.
1. The `branches`section contains the list of `branch protections` that need to be applied to all the repos in the org.
1. If the name of the branch is `default` in the settings, it is applied to the `default` branch of the repo.
1. `Validator`section to validate repo names using `regex`patterns

```yaml
# These settings are synced by https://github.com/github/safe-settings
# The `repositories` section contains repositories that need to be configured explicitly
#
# The `labels` section 

repositories: []
 
# The following attributes are applied to any repo within the org
# So if a repo is not listed above is created or edited
# The app will apply the following settings to it
labels:
  # Labels: define labels for Issues and Pull Requests
  - name: bug
    color: CC0000
    description: An issue with the system

  - name: feature
    # If including a `#`, make sure to wrap it with quotes!
    color: '#336699'
    description: New functionality.

  - name: first-timers-only
    # include the old name to rename an existing label
    oldname: Help Wanted
    color: '#326699'

>   - name: new-label

Reviewer: ```suggestion

  - name: new-label
```

- username: regpaco
  permission: push
# The permission to grant the collaborator. Can be one of:
# * `pull` - can pull, but not push to or administer this repository.
# * `push` - can pull and push, but not administer this repository.
# * `admin` - can pull, push and administer this repository.
- username: beetlejuice
  permission: pull
  exclude:
  - actions-demo
# You can exclude a list of repos for this collaborator and all repos except these repos would have this collaborator
- username: thor
  permission: push
  include:
  - actions-demo
  - another-repo
# You can include a list of repos for this collaborator and only those repos would have this collaborator

# See https://developer.github.com/v3/teams/#add-or-update-team-repository for available options
teams:
  - name: core
    # The permission to grant the team. Can be one of:
    # * `pull` - can pull, but not push to or administer this repository.
    # * `push` - can pull and push, but not administer this repository.
    # * `admin` - can pull, push and administer this repository.
    permission: admin
  - name: docss
    permission: push
  - name: docs
    permission: pull

branches:

>   # If the name of the branch is default, it will create a branch protection for the default branch in the repo

Reviewer: ```suggestion
  # If the name of the branch value is specified as `default`, then the app will create a branch protection rule to apply against the default branch in the repo
```

- username: regpaco
  permission: push
# The permission to grant the collaborator. Can be one of:
# * `pull` - can pull, but not push to or administer this repository.
# * `push` - can pull and push, but not administer this repository.
# * `admin` - can pull, push and administer this repository.
- username: beetlejuice
  permission: pull
  exclude:
  - actions-demo
# You can exclude a list of repos for this collaborator and all repos except these repos would have this collaborator
- username: thor
  permission: push
  include:
  - actions-demo
  - another-repo
# You can include a list of repos for this collaborator and only those repos would have this collaborator

# See https://developer.github.com/v3/teams/#add-or-update-team-repository for available options
teams:
  - name: core
    # The permission to grant the team. Can be one of:
    # * `pull` - can pull, but not push to or administer this repository.
    # * `push` - can pull and push, but not administer this repository.
    # * `admin` - can pull, push and administer this repository.
    permission: admin
  - name: docss
    permission: push
  - name: docs
    permission: pull

branches:

>   # If the name of the branch is default, it will create a branch protection for the default branch in the repo

Reviewer: ```suggestion
  # If the name of the branch value is specified as `default`, then the app will create a branch protection rule to apply against the default branch in the repo
```

Author: Sounds good. I will commit your changes.
